File: python/0217. Contains Duplicate.py
'''
Obvious implementation
Time complexity: O(N^2)
Space complexity: O(1)
'''
class Solution(object):
    def containsDuplicate(self, nums):
        for idx in range(len(nums)):
            for idx2 in range(idx+1, len(nums)):
                if nums[idx] == nums[idx2]:
                    return True
        return False    


'''
Efficient implementation
Time complexity: O(N)
Space complexity: O(N)
'''
class Solution(object):
    def containsDuplicate(self, nums):
        storage = set()        
        for num in nums:
            if num in storage:
                return True
            else:
                storage.add(num)
        return False      

File: java/0217. Contains Duplicate.java
package leetcode.java;

import java.util.HashSet;

/**
 * Obvious implementation
 * Time complexity: O(N^2)
 * Space complexity: O(1)   
 */
class Solution {
    public static boolean containsDuplicate(int[] nums) {
        for (int i=0; i<nums.length; i++) {
            for (int j=i+1; j<nums.length; j++) {
                if (nums[i] == nums[j]) {
                    return true;
                }
            }
        }
        return false;
    }
}



/**
 * Efficient implementation
 * Time complexity: O(N)
 * Space complexity: O(N)   
 */
class Solution {
    public boolean containsDuplicate(int[] nums) {
        HashSet<Integer> set = new HashSet<>();

        for (int i : nums) {
            if (set.contains(i)) {
                return true;
            } else {
                set.add(i);
            }
        }
        return false;
    }
}